state-sync/storage-service/server/src/lib.rs

```diff
@@ -446,26 +446,49 @@ fn get_peers_with_ready_subscriptions<T: StorageReaderInterface>(
 
     // Identify the peers with ready subscriptions
     let mut ready_subscriptions = vec![];
+    let mut invalid_peer_subscriptions = vec![];
     for (peer, data_subscription) in data_subscriptions.lock().iter() {
-        if data_subscription.highest_known_version() < highest_synced_version {
+        let highest_known_version = data_subscription.highest_known_version();
+        if highest_known_version < highest_synced_version {
             let highest_known_epoch = data_subscription.highest_known_epoch();
-            let target_ledger_info = if highest_known_epoch < highest_synced_epoch {
+            if highest_known_epoch < highest_synced_epoch {
                 // The peer needs to sync to their epoch ending ledger info
-                get_epoch_ending_ledger_info(
+                let epoch_ending_ledger_info = get_epoch_ending_ledger_info(
                     cached_storage_server_summary.clone(),
                     data_subscriptions.clone(),
                     highest_known_epoch,
                     lru_storage_cache.clone(),
                     data_subscription.protocol,
                     storage.clone(),
                     time_service.clone(),
-                )?
+                )?;
+
+                // Check that we haven't been sent an invalid subscription request
+                // (i.e., a request that does not respect an epoch boundary).
+                if epoch_ending_ledger_info.ledger_info().version() <= highest_known_version {
+                    invalid_peer_subscriptions.push(*peer);
+                } else {
+                    ready_subscriptions.push((*peer, epoch_ending_ledger_info));
+                }
             } else {
-                highest_synced_ledger_info.clone()
+                ready_subscriptions.push((*peer, highest_synced_ledger_info.clone()));
             };
-            ready_subscriptions.push((*peer, target_ledger_info));
         }
     }
+
+    // Remove the invalid subscriptions
+    for peer in invalid_peer_subscriptions {
```

Comment:
will subscriptions be recreated after epoch change?

Reply:
Yes, clients will always re-request subscriptions or data after a timeout 😄